### PR TITLE
Rotary dimmer

### LINKED ISSUE #8: Support for LED Dimmer Z-Wave 200W
Is it possible to add support for
https://www.elektroimportoren.no/led-dimmer-z-wave-200w/1402756/Product.html

### PR DESCRIPTION
**Pull request by fiLLLip**
Fixes #8

### CODE REVIEW

File: drivers/ZW_Dimmer_Rotary/device.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		this.registerCapability('dim', 'SWITCH_MULTILEVEL');
+
+		this.registerReportListener('BASIC', 'BASIC_REPORT', (report) => {
+			if (report && report.hasOwnProperty('Current Value')) {
+				if (this.hasCapability('onoff')) this.setCapabilityValue('onoff', report['Current Value'] > 0);
+				if (this.hasCapability('dim')) this.setCapabilityValue('dim', report['Current Value'] / 99);
+			}
+		});
+
+		// this.registerSetting('dim_duration', value => new Buffer([value === -1 ? 255 : util.calculateZwaveDimDuration(value)]));
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/device.js]: What exactly does this line do? It's commented out in all/most of the drivers. I suggest the line is removed.

File: drivers/ZW_Dimmer_Rotary/assets/icon.svg [hunk 2]
```diff
@@ -0,0 +1,810 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/assets/icon.svg]: The icon should in my opinion show the rotary dimmer (that the user actually sees), not the back panel from the user's manual. I'll have a stab at making an SVG for that if I'm unable to find a suitable one from one of the other Homey apps repos.

File: drivers/ZW_Dimmer_Rotary/driver.compose.json [hunk 3]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/driver.compose.json:1]: Not sure if I've understood the Z-wave parameters and their way of working properly, but looking at the Homey app and the [user manual](https://www.elektroimportoren.no/docs/lib/1402756-Brukerveiledning-5.pdf), there is a mismatch between the parameters and their meaning.

Parameter | Dimmer description | Homey device settings
----------|--------------------|------------------
1 | State after power restored | _Not listed/found_
2 | Notification when load status change | Restore state after power failure
3 | Default dimming speed (0-100 seconds) | Send basic report
4 | _Not listed_ | Dimmer duration (0-127 seconds)
5 | Enable/disable external switch to pair network | Minimum brightness
6 | Dimming way (linear or logarithmic) | MOSFET driving type
7 | Default brightness | Enable external switch
8 | Setting delay time when turn off the load, unit in second (0-60) | _Not found_

Ideally, this is aligned, right? Looking at the code in this repo, I'm guessing most of the parameters and their meaning are inherited (from homey-meshdriver?). My limited knowledge, i'm not able to figure out (yet?) how to change this for the rotary dimmer.

File: drivers/ZW_Dimmer_Rotary/assets/learnmode.svg [hunk 4]
```diff
@@ -0,0 +1,810 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/assets/learnmode.svg]: I suggest removing some of the details in the picture to make it less noisy + increase text size of e.g. "Reset" to make it more visible. I have a proposal, but need to figure out how to share it over Github (do I commit to your branch, or do I make another pull request onto feat/rotary-dimmer, @fiLLLip?).

File: drivers/ZW_Dimmer_Rotary/device.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+const { ZwaveDevice } = require('homey-meshdriver');
+
+const util = require('./../../node_modules/homey-meshdriver/lib/util');
+
+module.exports = class ZW_DimmerRotary extends ZwaveDevice {
+	onMeshInit() {
+
+		// enable debugging
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/device.js:10]: I suggest commenting out lines 11 and 14.

**Review by aasmundbo — CHANGES_REQUESTED** [review]
[drivers-ZW_Dimmer_Rotary-assets__svg.zip](https://github.com/fiLLLip/namron-homey/files/4209709/drivers-ZW_Dimmer_Rotary-assets__svg.zip)

File: drivers/ZW_Dimmer_Rotary/device.js (hunk 1 above)
Comment (fiLLLip) [reply drivers/ZW_Dimmer_Rotary/device.js]: It was a copy/pasta from the upstream repo, so I agree :) I guess it's used to set how long it should dim. Kind of a smooth "set level". "Use x seconds to reach level y"

File: drivers/ZW_Dimmer_Rotary/assets/icon.svg (hunk 2 above)
Comment (fiLLLip) [reply drivers/ZW_Dimmer_Rotary/assets/icon.svg]: Yeah, I just stole the one in the manual

File: drivers/ZW_Dimmer_Rotary/driver.compose.json (hunk 3 above)
Comment (aasmundbo) [reply drivers/ZW_Dimmer_Rotary/driver.compose.json:1]: >  I'm guessing most of the parameters and their meaning are inherited (from homey-meshdriver?)

I'm learning as I go here, so my last comment above is probably not correct. Sorry for the commenting spam.

Comment (fiLLLip) [reply drivers/ZW_Dimmer_Rotary/driver.compose.json:1]: Well, I have done absolutely no research on this myself  :D I just forked the upstream repo and changed some vendor/product ids. The rest is magic :)

File: drivers/ZW_Dimmer_Rotary/assets/learnmode.svg (hunk 4 above)
Comment (fiLLLip) [reply drivers/ZW_Dimmer_Rotary/assets/learnmode.svg]: I guess your second approach is the most clean

File: drivers/ZW_Dimmer_Rotary/device.js (hunk 5 above)
Comment (fiLLLip) [reply drivers/ZW_Dimmer_Rotary/device.js:10]: If I'm correct, this is in here for a purpose. It's difficult to debug remotely from crash reports having no info. Having at least some data would provide useful log statements. Homey users will never see this anyways?

File: drivers/ZW_Dimmer_Rotary/assets/learnmode.svg (hunk 4 above)
Comment (aasmundbo) [reply drivers/ZW_Dimmer_Rotary/assets/learnmode.svg]: I don't have access to push to the repo. I attached the 2 svg files to my review as a ZIP.

File: drivers/ZW_Dimmer_Rotary/device.js (hunk 5 above)
Comment (aasmundbo) [reply drivers/ZW_Dimmer_Rotary/device.js:10]: You're probably right 😄 . I just saw that it was commented out from some of the other devices in the repo.

File: drivers/ZW_Dimmer_Rotary/driver.settings.compose.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      "en": "Dimming",
+      "nl": "Dimmen"
+    },
+    "collapsed": true,
+    "children": [
+      {
+        "$extends": "dimDuration",
+        "$id": "dim_duration"
+      },
+      {
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/driver.settings.compose.json]: The rotary dimmer does not support configuring minimum brightness (it's configured by knobs on the dimmer itself).

File: drivers/ZW_Dimmer_Rotary/driver.settings.compose.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        "$extends": "saveState",
+        "$id": "save_state"
+      },
+      {
+        "$extends": "sendBasicReport",
+        "$id": "send_basic_report"
+      },
+
+      {
```
Comment (aasmundbo) [inline drivers/ZW_Dimmer_Rotary/driver.settings.compose.json]: The rotary dimmer does not support driving type (you cannot configure leading edge (GLI?) vs. falling edge trigger (GLE?)). The rotary dimmer is GLE only.

File: drivers/ZW_Dimmer_Rotary/driver.compose.json (hunk 3 above)
Comment (aasmundbo) [reply drivers/ZW_Dimmer_Rotary/driver.compose.json:1]: I have an idea how to get the homey app provide the correct setting options, but a quick test last night didn’t seem to have an effect on the dimmer, so I’m not at the finishing line. I don’t think it’s too important to get it in place before releasing another version of the Namron app, tho.

Now that I’m a contributor (thanks!) I can creat a pull request once I figure it out and get it working.

**Comment by fiLLLip** [comment]
@aasmundbo Do you have any other remarks, or is this ready to ship to test/prod?

**Comment by aasmundbo** [comment]
No other remarks from me. Please go ahead and merge.

As mentioned, I’ll try to make separate pull request once I’ve figured out and modified the configuration options of the dimmer.
